Rust Foundation Maintainer Fund (#3931)
rust-lang/rfcs: text/0000-rfmf-rust-foundation-maintainer-fund.md

Author: To be quite frank, I think many companies *WILL* and *SHOULD* choose to sponsor the maintainer fund -- at least, I hope they will! That's the whole point. The reality is that, typically, the alternative is not "doing the hard thing" of hiring a maintainer full-time -- it is doing nothing at all.

If a company wants to hire a full-time maintainer, that's great, we encourage that. But what I've found over the last 10+ years of working in open source is that really the incentives within a company are just not aligned for open-ended maintenance as a full-time role.

I myself am tired of tilting at windmills, and would rather take a stab at setting up a structure that provides the open-ended working environment directly.

(I don't think it's the only way to do open-source. For myself, I get a lot of value out of being at a company and having that need to show value. I think that's precisely what makes open-source great, that people can come and meet their own needs, so I like having a clear customer that I can work with to get a deep understanding of their needs, it gives me something to prioritize with, and that's useful. But not everybody likes that.)

Reviewer: I think the RFC is  very good. Thanks to everyone who worked on it.

I have two big questions.
- Will it actually work? I.e. will enough companies/individuals contribute to get a meaningful number of MiRs, in a way that persists?
- What's the pay rate for an MiR? The choice here will have a big impact on the effectiveness of the program. A lower number would allow for more MiRs, a higher number might entice more productive MiRs.

I think it's fine that the RFC doesn't answer these questions. I just mention them as the things that I am now wondering about.

Author: I added a FAQ that I think covers this (and links to the thread). Resolving. Thanks!

Selecting Maintainers in Residence is a collaboration between the Foundation and a "Funding team" appointed by the Leadership Council. This Funding team will weigh the set of applications against the project's needs and priorities.

The Funding team is additionally charged with ensuring the program's overall success. When sponsors contribute undirected funding, they are investing in the Rust project as a whole — and the project should meet them in good faith. Project teams receiving support from the program are expected to help the Funding team manage sponsor relations, e.g., by meeting with sponsors or providing other reasonable sponsor benefits.

[RFC 3919]: https://github.com/rust-lang/rfcs/pull/3919

This RFC was jointly written by the [RFMF Design Committee](https://github.com/rust-lang/team/blob/0acc660a6bce2b9166362d4bfcbed872508085a6/teams/rfmf-design-committee.toml).

# Motivation
[motivation]: #motivation

> The Rust Foundation is establishing a Maintainer Fund to collect sponsorships and provide long-term funding for Rust maintenance. Funds raised through the RFMF are earmarked to fund Rust maintainers under the supervision of the Leadership Council. The Leadership Council commissioned this RFC to recommend how those funds should be used.

Reviewer 2: Some common term in not-for-profits are:
- dedicate(d)
- direct(ed)
- restrict(ed)

"Restricted" is the most common legal term, but I think "dedicated" is more readable.

We recommend that the funds be directed into the Project Priorities budget with an earmark that they should be used only for programs that directly fund individual Rust maintainers (which includes project management). Earmarking the funds helps ensure that the "sales pitch" is clear: donations given to the RFMF will go directly into a maintainer's pocket.

We further recommend that the Leadership Council create a "Maintainer in Residence" program to augment the existing project management program and the grants program proposed in [RFC 3919]. Maintainers in Residence are maintainers who are paid to work on a full-time or substantial part-time basis to maintain some part of the project.

> ## Why focus on long-term maintenance?

Reviewer 3: Also, just thinking about this more, but this this header is actually just what the entire motivation section should be: we have this experience from the past and know what maintainers are saying, and that's what motivates the entire RFC. Details on the specifics are stuff that should go under the sections detailing the specifics, not the motivation.

> ## Why focus on long-term maintenance?

Author: I reworked the section headings; I generally prefer section headings that summarize the "main point(s)" that follows, not sure why I went for the question style here.

Although the default is for the project to pick the area of focus for a MiR, we do allow the Funding team the latitude to offer involvement in area selection as a sponsor benefit at higher tiers. The intention is to permit a company that has a strong need on a particular area to fund a maintainer in that area, if they are willing.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

The RFMF collects sponsorships from companies and individuals. Funds support project grants, project management, and the Maintainer in Residence program. Maintainers in Residence are experienced, self-directed maintainers who do the work that keeps Rust healthy. They participate in team discussions, review PRs, mentor newcomers, and work on what the team needs.

## Design axioms

### Not one size fits all

Maintainers have a wide variety of needs and no one program will work for everyone. We allow RFMF funds to be used for any kind of program that directly benefits maintainers.

### The MiR is a collaboration between the Project and the Foundation

> Neither the Foundation nor the Project can operate the MiR program on their own. The Foundation has a bank account, legal entity, and operational capacity; the Project has knowledge of team health and needs. The Foundation is the incoming channel by which most sponsors arrive; the Project governs the codebase that sponsors want to support. This RFC proposes that both project members (the Funding team) and Foundation staff jointly make major decisions. This is a partnership, not a handoff.

Reviewer 4: Now there's a specific FAQ entry down this document ([What's the relationship between the Rust Project and the Rust Foundation?](https://github.com/rust-lang/rfcs/pull/3931/files#diff-7067f2fbc93408343a6ef705a7691d2d2a56942c44c70f866f067db889b5d9d7R258))

### What is not allowed as a benefit

*What sponsors do not get:* the ability to unilaterally direct a maintainer's work, pick who gets added to project teams, or otherwise bypass project processes.

## Selection process is driven by a team within the project, supported by Foundation staff

When funding is available, the Funding team and Foundation put out an open call for applications. The Funding team and Foundation staff review applications, consider the project's needs, and then the Foundation makes offers to the strongest candidates.

## The Funding team owns the project's long-term success

The Funding team owns the program's overall success. They keep up with teams to understand where support is needed and how well the program is working; they can adjust aspects of the program to make it work better over time.

The Leadership Council as well as the teams benefitting from the work of Maintainers in Residence are expected to support the Funding teams' efforts (e.g., by meeting with sponsors upon request or otherwise helping out with the sponsor benefits described above).

## What Maintainers in Residence do

> Maintainers in Residence split their time between team priorities and individual priorities of their own choosing within their area of focus. The exact balance varies depending on the individual, their experience, and the needs of the team — the important thing is that both team-directed and self-directed work are expected. This is about "team-directed vs. self-directed," not "maintenance vs. features." The PSF's experience after nearly five years confirms that focusing purely on team-directed needs and multiplicative maintenance can be very draining; giving time for self-directed projects "made all the difference" in satisfaction.

Reviewer 3: It makes more sense to me to explicitly link this point to the PSF section in Prior Art rather than just lightly summarising it here and not linking the two. There's a lot of document between the two points and this summary doesn't give enough information to really satisfy as a justification.

Either that, or potentially the PSF prior art could be moved to the motivation section as an explicit motivating feature, rather than saving it for a note at the end.

Author: Added a citation.

The Funding team prioritizes applications based on:

1. conversations with team leads and team members to assess what support is most urgently needed;
2. the applicant's history with the project;
3. any specific work that was proposed in the application;
4. the applicant's availability and whether it suffices for the tasks they expect to perform;
5. the results of interviews or conversations with the applicant; and
6. any history of interpersonal or Code-of-Conduct issues known to the moderation team.

The Funding team works with the Foundation to select from the applicant pool and to extend offers. The Funding team is looking for maintainers that have technical depth in the relevant area, community standing and trust within the Project, and sustained work orientation (a track record that suggests they'll thrive in a role focused on reviews, mentorship, unblocking, and the kind of long-term technical work that requires deep context).

## Contract terms

> The precise terms of the working arrangement are not defined by this RFC. The important points are: it is a **substantial commitment** (full-time or significant part-time), and it is expected to **continue (or be renewed) as long as both sides are satisfied and funding is available**. Some areas may not require a full person's time; it is fine to have one person cover two areas, or two people each contribute part-time to a single area, so long as there is enough concentrated time to build and maintain deep context.

Reviewer 5: Also note that several of us have explicitly said that we would like to see employment-based opportunities available as well, and that information has been conveyed to Foundation management.

> The precise terms of the working arrangement are not defined by this RFC. The important points are: it is a **substantial commitment** (full-time or significant part-time), and it is expected to **continue (or be renewed) as long as both sides are satisfied and funding is available**. Some areas may not require a full person's time; it is fine to have one person cover two areas, or two people each contribute part-time to a single area, so long as there is enough concentrated time to build and maintain deep context.

Reviewer 6: That's fair, but needs to be clarified more in the RFC. It keeps using the word "contract" without an explicit disclaimer and also talks about "renewing the contract" which is not an explicit thing with employment.

## Flat pay structure

> Compensation should be a flat rate within a small number of bands (e.g., junior and senior), rather than individually negotiated. Flat rates keep the program simple and avoid the perception that some maintainers are valued more than others for equivalent work. The compensation structure is publicly advertised as part of the call for applications. The Foundation determines the specific rates and may adjust them over time as the program learns what works and what's needed to attract strong candidates.

Reviewer 2: If we go down the path of considering expenses, then a major expense that varies between jurisdictions is tax. That's on top of (or before) the cost of housing, healthcare, and similar required expenses is taken out.

And paying someone less, just because they live in a particular area, means they're stuck there. It also doesn't cover people with resident rights in multiple countries. Some people even move between jurisdictions during the year.

Analysing expenses is also a privacy risk: the org will hold significant personal information on staff, beyond what's needed to administer their day to day jobs. 

There's also a wider drawback with negotiating pay rates: you select for people with pay negotiation skills, rather than project-relevant skills.

Maybe it's worth trialling a simple flat rate to start with, then reviewing after a year or two?

Maintainers in Residence are also expected to:

* respond to reasonable requests from the Funding team on behalf of sponsors;
* keep records of their activity as needed by the Funding team (see below);
* resolve time-critical issues such as urgent bugs;
* champion Project Goals supported by their respective teams, even if they themselves might not have championed that goal as an individual;
* work with the Project to ensure their work gets regularly reported on;
* remain a member of the Project and relevant teams, in good standing.

## Reporting and impact visibility

The Funding team may request that MiR collect data on their activity so that they can prepare impact reports and other material for sponsors. Whenever possible, however, the Funding team should work with other teams (e.g., the Goals or Content teams) to handle the creation of that material, so that the MiR can focus primarily on maintenance.

## The Funding team collects feedback (positive or negative) on MiR performance

> Feedback on MiR performance (positive or negative) can be directed to the Funding team. The Funding team will also periodically seek feedback from the project proactively. The Funding team and the manager at the Foundation will convey feedback to the MiR at regular intervals. 

Reviewer 2: Is it worth saying that feedback is initially private?

```suggestion
Feedback on MiR performance (positive or negative) can be directed privately to the Funding team. The Funding team will also periodically seek feedback from the project proactively. The Funding team and the manager at the Foundation will convey feedback to the MiR at regular intervals. 
```

## The Manager works with the Funding team to communicate feedback

The *manager* for a MiR is charged with making sure they are well taken care of. They should meet regularly with the MiR, monitor their workload to ensure that it is balanced and they are not being given either too many or too few responsibilities. They should also have performance conversations at regular intervals to give the MiR a sense for whether they are doing well. Finally, if negative feedback has been received, they should have a clear conversation with MiR to set expectations.

If the program grows to large number of MiRs, however, we recommend that the Leadership Council use some portion of the RFMF funds to hire a dedicated manager who would work closely with the Funding team (see [Who manages Maintainers in Residence after they're hired?][who-manages-mirs]).

## Moderation concerns about a MiR

> Code of Conduct violations that result in removal from the Project immediately terminate the contract. Team membership is a hard requirement for the role; a maintainer who is no longer a project member cannot continue as a Maintainer in Residence.

Reviewer 7: We mods would feel very hesitant to immediately cut off someone's funding no matter what the reason is that we moderated them.

I would really like it if there were a guaranteed period that the MiR would keep being funded. Usual minimums (e.g. in Germany) are 1 month (but more if the employee is part of a company for several years).

Reviewer 8: I agree. I don't think it's within the scope of the RFC to specify concrete terms here though. Since it overlaps with legal requirements it's really in the scope of the Foundation. What the RFC can do is provide guidelines here, like

> Where possible, contracts should include a termination period to ease the transition for anyone whose contract must be terminated.

It can also suggest certain minimums, as you do, but ultimately it would be up to the Foundation to decide.

Reviewer 5: It's not within the scope of the RFC to specify the *legal* terms; the only part that matters for the purposes of the RFC is "the *role as a Maintainer in Residence* ends immediately", as they can no longer perform that role (they are not a team/project member). If there is some obligation to keep paying them (e.g. local employment/contract law), that's a different matter, but does not mean they continue on *as a Maintainer in Residence* past that point. "Fired, but we're obligated to handle financial obligations X and Y" is a well-understood state for an employer/client to deal with.

Reviewer 3: I think this is also relevant to the discussion I mentioned in the "contract renewal" section: https://github.com/rust-lang/rfcs/pull/3931#discussion_r2944036403

Essentially: short of situations like the MIR suddenly committing a crime or spontaneously revealing that they were secretly a disgusting racist the entire time, removal should not be a surprise to anyone involved. The main point here is probably that the code of conduct is another set of rules the MIR is expected to follow and taken as seriously as something like performance, and not that they intend to trigger like a trap and cartoonishly yoink the maintainer out of the project in a net hanging from a tree branch.

However, teams should also bear in mind that MiR do not have infinite capacity. MiR can feel free to decline to work on team priorities past a certain point. The MiR manager and Funding team will help to negotiate this balance as needed.

# Frequently asked questions
[faq]: #faq

## What does it look like to have a Maintainer in Residence on my team?

In one sense, the same as having any other team member. They show up in reviews, participate in design discussions, mentor newcomers, and work on what the team needs. The PSF's experience after nearly five years is that the Developer in Residence does roughly 50/50 maintenance and proactive work, and the role feels like "just another team member" rather than an outside presence.

Unlike a volunteer, however, MiR are expected to put some portion of their time towards team priorities. Whereas before, if there was an urgent task, a team had to hope that somebody had the capacity to pick it up, they can now direct sustained effort where it's needed most. This is a new capability.

## Why not a flexible contractor pool instead of long-term maintainers?

> Context and trust take time to build. The maintenance problems we heard about in team lead interviews (review backlogs, cross-team blocking, complex refactors that need months of sustained attention) require sustained presence, not project-scoped interventions. Contractors for scoped work is a valid model, but it's a different program solving a different problem.

Author: I'm ok with the language as is.

## Why are RFMF funds earmarked for maintainers?

Sponsors contributing to the RFMF want to know their money is going directly to fund maintainers. Earmarking gives the fund a clear value proposition: every dollar goes to paying people who maintain Rust. Without earmarking, the Leadership Council could use RFMF funds for any purpose — travel grants, event sponsorship, infrastructure — all valuable, but harder to explain to a company evaluating its return on investment. Since money is fungible, earmarking RFMF funds for maintainers also frees up other budget for those other purposes.

The earmark is broad: Maintainers in Residence, project grants, and the program management overhead needed to support them. The Leadership Council determines the specific form. This gives the Council real flexibility while keeping the fund's purpose legible to sponsors.

## What does this RFC deliberately not specify?

This RFC does not define how the RFMF collects sponsorships, the precise tiers or benefits sponsors receive, the precise pay structure, or other fine-grained details. The Funding team and Foundation can work these out.

> # Prior art

Reviewer 3: Since this applies to multiple examples here: it would be great if each of the examples mentioned directly linked to articles discussing the programs in question and their experiences, rather than simply summarising them, so that folks who want to learn more about specific aspects can do their own research more easily.

## Project Grants Program: a related committee model

The Project Grants Program ([RFC 3919]) proposes a program supporting a handful of contributors with modest monthly stipends. It charters a Grants team (5 members, LC-appointed, organized as a Launching Pad subteam) to select recipients and oversee the program. The RFC explicitly positions itself as "distinct from, but complementary to" the MiR: grants are smaller-scale, flexible, Project-controlled support, while the RFMF targets larger-scale, sustained maintenance.

The Grants team's charter overlaps significantly with the Funding team charter we define here — both involve assessing project needs and selecting candidates. The Leadership Council may choose to extend the Grants team with the Funding team's responsibilities rather than creating a separate body, which would avoid fragmenting the Project's attention across multiple committees with overlapping mandates.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

> ## Organizational form of the Funding team

Reviewer 3: An immediate obvious response to this section is that people can be on multiple teams.

I don't think that it's really an unresolved question at this point, rather, an unresolvable question: it's not going to be clear whether two teams is the ideal choice until there are more people involved and more decisions are made. It seems reasonable to just keep them combined into a general "financial" team for now and allow them to specialise into subteams if and when the time warrants it.

Obviously, it feels bold to have an empty "unresolved questions" section but I don't really think this one fits here, and I'm sure that plenty of questions, even frequently asked ones, can be placed here.

- Feature Name: `rust_foundation_maintainer_fund`
- Start Date: 2026-02-23
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)

# Summary
[summary]: #summary

> This RFC defines the relationship between the Rust Foundation Maintainer Fund (RFMF) and the open-source Rust project. The RFMF is a dedicated fund used to support Rust maintenance: open-ended, multiplicative work that improves Rust and its codebase and makes it more accessible.

Reviewer: I'm not sure what "more accessible" means here, nor if it's referring to "Rust" or "its codebase" or both. I feel like removing "and makes it more accessible" would be clearer and wouldn't lose anything.

Author: ```suggestion
This RFC defines the relationship between the Rust Foundation Maintainer Fund (RFMF) and the open-source Rust project. The RFMF is a dedicated fund used to support Rust maintenance: open-ended, multiplicative work that improves Rust and its codebase.
```

Reviewer 5: (EDIT: Somehow GitHub did not show me Niko's comment when I responded to this.)

## Selection process is driven by a team within the project, supported by Foundation staff

When funding is available, the Funding team and Foundation put out an open call for applications. Where appropriate, the Funding team may also proactively reach out to potential applicants to encourage them to apply, if they may be a good fit for areas the project needs. The Funding team and Foundation staff review applications, consider the project's needs, and then the Foundation makes offers to the strongest candidates.

## The Funding team owns the project's long-term success

The Funding team owns the program's overall success. They keep up with teams to understand where support is needed and how well the program is working; they can adjust aspects of the program to make it work better over time.

The Leadership Council as well as the teams benefitting from the work of Maintainers in Residence are expected to support the Funding teams' efforts (e.g., by meeting with sponsors upon request or otherwise helping out with the sponsor benefits described above).

## What Maintainers in Residence do

> Maintainers in Residence split their time between team priorities and individual priorities of their own choosing within their area of focus. The exact balance varies depending on the individual, their experience, and the needs of the team — the important thing is that both team-directed and self-directed work are expected. This is about "team-directed vs. self-directed," not "maintenance vs. features." The PSF's experience after nearly five years confirms that focusing purely on team-directed needs and multiplicative maintenance can be very draining; giving time for self-directed projects "made all the difference" in satisfaction (see [Prior Art][prior-art]).

Reviewer: This is the first mention of PSF but the abbreviation isn't defined.

Author: ```suggestion
Maintainers in Residence split their time between team priorities and individual priorities of their own choosing within their area of focus. The exact balance varies depending on the individual, their experience, and the needs of the team — the important thing is that both team-directed and self-directed work are expected. This is about "team-directed vs. self-directed," not "maintenance vs. features." The Python Software Foundation's experience after nearly five years confirms that focusing purely on team-directed needs and multiplicative maintenance can be very draining; giving time for self-directed projects "made all the difference" in satisfaction (see [Prior Art][prior-art]).
```

## The Funding team's day-to-day responsibilities

The Funding team's role is to keep a pulse on the project and work with the Foundation to select which maintainers to hire. Its core responsibilities are:

1. **Staying in regular contact with teams** — meeting regularly with team leads and members to understand their needs, health, and where support would have the most impact.
    * These meetings can be helpful for purposes other than hiring a MiR: the Funding team may be able to connect the team with Foundation resources to resolve a situation.
2. **Working with the Foundation to select MiR candidates** — when positions become available, evaluating applicants and selecting candidates who'll have the most overall impact based on project needs.
3. **Collecting feedback on how well the program and the MiRs are working** — as the team responsible for selecting which maintainers to hire, the Funding team is also the team responsible for fielding feedback on how well those decisions work out and making adjustments as needed.

The Foundation supports the Funding team with logistics. The Foundation issues contracts or manages employment.

> 

Reviewer 5: *[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3931/c143e315774f746d667c5eecd95a8ed999e8a729..45ee79dda91d4f0303444e389767a0933e2e02d3)*

```suggestion

The Leadership Council appoints new members for the Funding team, for fixed terms (e.g. one year at a time), which can be renewed. This helps ensure that the membership reflects the will and needs of broader Project teams, since the Funding team should consider the needs of all of those teams. The charter of the Funding team should capture this.

```

## Working arrangements should be substantial and stable

The precise terms of the working arrangement are not defined by this RFC. Whether achieved through contracts, employment, or other means, the goal is to create a stable environment that allows maintainers to focus on their work in a sustained, year-over-year fashion. The important points are that the arrangement is substantial and stable:

* By *substantial*, we mean that the compensated time is either full-time or significant part-time. Some areas may not require a full person's time; it is fine to have one person cover two areas, or two people each contribute part-time to a single area, so long as there is enough concentrated time to build and maintain deep context. This latitude also allows us to accommodate maintainers who are not interested in a full-time role.
* By *stable*, we mean that it is expected to continue (or be renewed) as long as both sides are satisfied and funding is available.

## We recommend a flat pay structure to start

> We recommend starting with a single flat rate rather than individually negotiated compensation. A flat rate keeps the program simple, avoids the perception that some maintainers are valued more than others for equivalent work, and removes the disadvantage that negotiation creates for people who tend to undersell themselves. The rate should be publicly advertised as part of the call for applications.

Reviewer 5: *[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3931/c143e315774f746d667c5eecd95a8ed999e8a729..45ee79dda91d4f0303444e389767a0933e2e02d3)*

```suggestion
We recommend starting with a single flat rate or small fixed set of flat rates (e.g. 2-3) rather than individually negotiated compensation. Flat rates keep the program simple, avoids the perception that some maintainers are valued more than others for equivalent work, and removes the disadvantage that individual negotiation creates for people who tend to undersell themselves. The rates should be publicly advertised as part of any call for applications.
```

In the spirit of other parts of the RFC, we should give a bit of a "palette of choices" for the creation of this, rather than strictly limiting it. A small fixed set of rates is still in the spirit of the underlying goal here: avoiding the disadvantages of individual negotiation and the result in which people are paid differently for equivalent work based on how well they could negotiate. And a small fixed set of rates may allow us to successfully employ people at a wide range of skillsets whose work may *not* be equivalent.

This suggested change does not prescribe which approach we take, and just makes it possible to consider and adopt such approaches without going against the recommendations of this RFC.

* work with the Project to ensure their work gets regularly reported on;
* remain a member of the Project and relevant teams, in good standing.

## Reporting and impact visibility

The Funding team may request that MiR collect data on their activity so that they can prepare impact reports and other material for sponsors. Whenever possible, however, the Funding team should work with other teams (e.g., the Goals or Content teams) to handle the creation of that material, so that the MiR can focus primarily on maintenance.

The expectation is that a typical MiR could satisfy these requirements by registering project goals for their major initiatives, posting regular updates, and periodic meetings with their manager. Other teams would supplement this work. For example, the Content team may opt to interview a MiR and prepare a blog post covering their work, and the Funding team might gather Github activity automatically to quantify things like PR reviews or issue triage.

## The Funding team collects feedback (positive or negative) on MiR performance

> Feedback on MiR performance (positive or negative) can be directed privately to the Funding team. The Funding team will also periodically seek feedback from the project proactively. The Funding team and the manager at the Foundation will convey feedback to the MiR at regular intervals. Feedback, particularly negative feeback, is private and should not be shared publicly without permission from the MiR.

Reviewer: ```suggestion
Feedback on MiR performance (positive or negative) can be directed privately to the Funding team. The Funding team will also periodically seek feedback from the project proactively. The Funding team and the manager at the Foundation will convey feedback to the MiR at regular intervals. Feedback, particularly negative feedback, is private and should not be shared publicly without permission from the MiR.
```

**Project manages.** The Project has deeper technical context to evaluate whether work is landing well. But Project-side management means either a volunteer committee handles it (likely unskilled at management, outside the Funding team's competency) or a dedicated person is hired for it (high overhead relative to the program size, especially early on).

In practice, we expect this to be a phased approach. While the program has relatively few maintainers, the Foundation provides management skill and the Funding team provides feedback as an input. As the program grows, a dedicated support role may emerge — someone who meets with MiRs regularly, helps them build a portfolio of their work, and serves as the point of contact when teams have concerns. Whether that role lives within the Foundation, the Project, or somewhere in between is a question that becomes more important at scale and can be revisited as the program learns what it needs.

## What if a Maintainer in Residence underperforms?

See the ["Who manages MiR"][who-manages-mirs] question.

## What about people who only want to work part time?

> Maintainers in Residence can work substantial part-time — the key requirement is enough concentrated time to build and maintain deep context, not necessarily a 40-hour week. Some areas may not need a full person's time, and it's fine to have one person cover two areas or two people each contribute part-time to a single area. For contributors who want lighter-touch support, the LC's Project Grants Program ([RFC 3919]) is designed for exactly that. The two programs are complementary: grants support a broad base of contributors; the RFMF funds sustained maintenance work from people with deep context.

Author: I'd say...the RFC doesn't rule it out. I think that a full-time job at another company would be inadvisable, but a part-time job may make stance. I don't know how to draw a *hard line* about the number of hours we expect. I personally would prefer 3 days a week or more, but I feel comfortable deferring the details to the team.

## What about sponsors who want to pay for a particular item to get done?

> That's outside the scope of the RFMF, which is undirected funding — sponsors contribute to a general fund and the Leadership Council decides how to deploy it. A companion effort, the proposed Project Goals Funding program (see [Future possibilities][]), is designed for exactly this: sponsors direct funding at specific Project Goals, and the Foundation issues grants to advance that work. Sponsors seeking that level of direction should look there rather than the RFMF.

Reviewer 9: *[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3931/c143e315774f746d667c5eecd95a8ed999e8a729..45ee79dda91d4f0303444e389767a0933e2e02d3)*

```suggestion
That's outside the scope of the RFMF, which is undirected funding — sponsors contribute to a general fund and the Leadership Council decides how to deploy it. A companion effort, the proposed Project Goals Funding program (see [Future possibilities][#future-possibilities]), is designed for exactly this: sponsors direct funding at specific Project Goals, and the Foundation issues grants to advance that work. Sponsors seeking that level of direction should look there rather than the RFMF.
```

# Summary
[summary]: #summary

This RFC defines the relationship between the Rust Foundation Maintainer Fund (RFMF) and the open-source Rust project. The RFMF is a dedicated fund used to support Rust maintenance: open-ended, multiplicative work that improves Rust and its codebase.

The Leadership Council has a Project Priorities budget, which is used to fund various initiatives, such as travel grants or program management. RFMF funds will be directed to this budget, but they will be dedicated to activities that direct funding to individual maintainers. This includes the existing program management program and would include the proposed Project Grants Program ([RFC 3919]), which provides modest stipends to recognize and support existing contributors. It will also include a third program, the *Maintainer in Residence program*, proposed by this RFC.

The Maintainer in Residence program is dedicated to hiring long-term maintainers and funding their maintenance work in full. Maintainers' in Residence time is split between priorities guided by the teams they are supporting and priorities of their own choosing within the project.

> Selecting Maintainers in Residence is a collaboration between the Foundation and a "Funding team" appointed by the Leadership Council. This Funding team will weigh the set of applications against the project's needs and priorities.

Reviewer 10: *[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3931/c143e315774f746d667c5eecd95a8ed999e8a729..3e275ec13729e14088c69c8915c80f45cfa7ede3)*

```suggestion
Selecting Maintainers in Residence is a collaboration between the Foundation and a "Funding team" whose members are appointed by the Leadership Council or selected directly by the top-level teams. This Funding team will weigh the set of applications against the project's needs and priorities.
```

For consistency.

Reviewer 10: This was for consistency with an earlier-applied edit.  I've pushed a commit doing this to the branch so as to resolve this review comment.

> 

Reviewer 5: ```suggestion

Funding team members must not disproportionately come from any one company, legal entity, or closely related set of legal entities, to avoid impropriety or the appearance of impropriety. If the Funding team has 5 or fewer representatives, no more than 1 representative may have any given affiliation; if the Funding has 6 or more representatives, no more than 2 representatives may have any given affiliation. (This affiliation limit comes from [RFC 3392](https://rust-lang.github.io/rfcs/3392-leadership-council.html#limits-on-representatives-from-a-single-companyentity), and the Funding team charter should import the definition of affiliations and the processes around handling them from there. The Leadership Council may update this policy through its normal decision-making process.)

```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3931/c143e315774f746d667c5eecd95a8ed999e8a729..3e275ec13729e14088c69c8915c80f45cfa7ede3)*

Reviewer 11: Yeah, it seems very reasonable to have affiliaiton limits for this team. Good catch. I'm not sure if we should define the exact rules in this RFC, or just say that there should be limits and leave the details for later.

Reviewer 8: I would prefer not to spell them out in the RFC. My experience has been that these limits have been overly restrictive in the past.

To make this more concrete: Let's say the funding team consists of 5 members, and 2 of the nominees work in different parts of the same large company. What is the kind of impropriety we are trying to prevent in this situation? (As sometimes happens, we may struggle to get even 5 nominees that we feel are qualified, despite our efforts to recruit.)

Reviewer 2: Based on my experience of grants teams in other communities, the most frequent concern is (potentially) biased decision-making. For example, a sub-group (appears to) select a maintainer that favours their goals, rather than using established goal-setting processes.

There doesn't have to be any actual bias. I've seen preventing, investigating, and addressing concerns take up a huge amount of volunteer time and attention. I've been on committees like that, and it's very draining.

The thresholds can be adjusted. But the overall aim is to look diverse and independent to a casual observer.

Separately, just a minor wording tweak to the last sentence in Josh's suggestion:
```suggestion

Funding team members must not disproportionately come from any one company, legal entity, or closely related set of legal entities, to avoid impropriety or the appearance of impropriety. If the Funding team has 5 or fewer representatives, no more than 1 representative may have any given affiliation; if the Funding has 6 or more representatives, no more than 2 representatives may have any given affiliation. (This affiliation limit comes from [RFC 3392](https://rust-lang.github.io/rfcs/3392-leadership-council.html#limits-on-representatives-from-a-single-companyentity), and the Funding team charter should import the definition of affiliations and the processes around handling them from there.)

```

Reviewer 5: @teor2345 Thanks for the typo fix; incorporated into my original suggestion.

@tmandry wrote:
> I would prefer not to spell them out in the RFC. My experience has been that these limits have been overly restrictive in the past.

Based on your second sentence, I very much think they *do* need to be spelled out in the RFC. I don't think we should relitigate these, and I grabbed the ones from the Council precisely because they're terms we're already familiar with.

> To make this more concrete: Let's say the funding team consists of 5 members, and 2 of the nominees work in different parts of the same large company. What is the kind of impropriety we are trying to prevent in this situation?

I'll second everything @teor2345 said. The impropriety, or appearance of impropriety, is precisely that the funding team is in the position of prioritizing the allocation of funds within the project and selecting/nominating candidates within the project, and could very easily prioritize those known to a particular company or those whose goals align with a company. This does not even have to be *intentional*, but simply a matter of familiarity. A set of team members with diverse affiliations avoids either doing that or raising the question of it.

Reviewer 2: Thinking about this some more, here's one scenario I expect will come up:
- a company donates to the fund
- staff of that company are on the funding team, which decides how to spend that donation
- someone is concerned about staff spending company donations (either inside the company or inside the Rust project)

Currently, the proposed mitigations are:
- most sponsors pool their funds (but this is not guaranteed)
- affiliation limits

Is there an existing conflict of interest policy that applies, or could be imported (perhaps with modifications?
Should affiliates also recuse themselves from decisions to spend un-pooled donations from their company?

(Recusing from pooled funds is impractical because tainting is long-term, and hard to remember. Some kind of "majority of this year's donations" check might be possible, but it's still complicated. Recusing from discussions is also tricky, because a lot of discussions happen async.)

Reviewer 8: > Based on your second sentence, I very much think they do need to be spelled out in the RFC. I don't think we should relitigate these, and I grabbed the ones from the Council precisely because they're terms we're already familiar with.

You're saying that _because_ I've seen them be overly restrictive, you want to keep them as-is and spell them out in the RFC. It's.. hard for me not to see that as a low-trust way of engaging with my perspective.

I should be explicit that I am saying this with my project member hat on, but also with the perspective of someone who works for a large company. To clarify something that you might be reading between the lines of my comment, these limits have _not_ been overly restrictive to any company I've ever worked for. In fact, I have never heard a company complain about representation limits. In my view, they have been overly restrictive _to the project_.

The instances that inform my thinking involve large companies (not one I work for), where two excellent project members happen to work for the same company on completely different teams, and are unlikely to collude. The policy is restrictive because it leaves no room for project leadership to evaluate risk and make a judgment call.

In short, I think treating every entity, no matter the size, as equal is overly simplistic and restrictive. We should absolutely have a policy here that prevents impropriety or the appearance of impropriety – _and_ we should look for ways to improve the policy and guidance over time.

Another way it can be overly restrictive is that the project may later decide it _wants_ the perspective of sponsors explicitly represented on the funding team. This is not the way I would set it up initially, because I'm not sure it's necessary or a good idea. But we discussed this possibility many times on the committee, so I would be surprised to preclude it directly in the RFC. I'm saying this part mostly with my project hat on – but also with my "internal company Rust sponsor" hat on – because I want to give the Council options that make funding more appealing.

Concretely I propose setting an initial policy, but giving the Council latitude to update that policy over time. That can be accomplished by adding a sentence like:

> The Leadership Council may update this policy through its normal decision-making process.

Reviewer 5: > You're saying that because I've seen them be overly restrictive, you want to keep them as-is and spell them out in the RFC. It's.. hard for me not to see that as a low-trust way of engaging with my perspective.

That was not at all my intention here, and my apologies for coming across that way.

My reaction was entirely one of remembering the *extensive* discussions on this exact point during the design of the Council, and not wanting to replay those same arguments in substantially similar form. That doesn't preclude *having the conversation*. My goal in spelling out the *initial* policy in the RFC was specifically to have something to *start* with that we have precedent for and have operated with for a while now, in order to make forward progress.

> To clarify something that you might be reading between the lines of my comment, these limits have not been overly restrictive to any company I've ever worked for.

I was not reading anything between the lines there; again, sorry for the implication.

> The instances that inform my thinking involve large companies (not one I work for), where two excellent project members happen to work for the same company on completely different teams, and are unlikely to collude. The policy is restrictive because it leaves no room for project leadership to evaluate risk and make a judgment call.

Part of the point is that bad things can happen *even when people aren't colluding or otherwise acting in bad faith*, because working for the same company and modeling their thinking and priorities day in and day out will tend to affect your thinking, not necessarily intentionally. So, several people working for the same company can end up bringing similar mindsets or priorities to bear without ever having coordinated with each other.

I think when people see affiliation limits and conflict of interest policies like this, the natural inclination is to assume that they're guards against *malice*; however, I would expect the vast majority of the time they avert unintentional problems.

> Another way it can be overly restrictive is that the project may later decide it wants the perspective of sponsors explicitly represented on the funding team.

> I would be surprised to preclude it directly in the RFC.

I would not expect this language to preclude that in any way; my expectation would be that if we have some process to, for instance, add a representative of sponsors above a given threshold, that wouldn't run afoul of this limit. We could even, at that time, decide to state that the official sponsor representatives didn't count towards the limit, if we thought that would be an issue. (I wouldn't expect it to be, though, in practice.)

> Concretely I propose setting an initial policy, but giving the Council latitude to update that policy over time. That can be accomplished by adding a sentence like:

>    The Leadership Council may update this policy through its normal decision-making process.

I would already have assumed it to be the case that the Council could update the policy (and anything else about this), but I've added that sentence to this suggestion, to make that explicitly clear. That seems like a good idea.

Reviewer 8: Thanks for clarifying and for the addition.

To be honest, I still worry that the 1-person limit is overly restrictive when we are bootstrapping the team. I expect that a small team will be more effective, and the most likely outcome is that the needs of the project are well represented – the members are probably going to be existing project members, after all – while the needs of sponsors are under-represented. To me it makes sense to be intentional about representing the needs of funders in the beginning, when there are few and we are trying to get more, but without creating an expectation that they will always have the same kind of influence. The affiliation limit makes this more difficult; to date I only know of 2 major funders, which limits the pool of potential members who can represent sponsor interests.

This seems like a different case from the leadership council, which is purely about representing project interests, and where the affiliation limit makes more sense to me. In contrast, I see the funding team as putting together project interests with sponsor interests, for the good of both (but ultimately always for the good of the project).

With that said, since the text says the policy can be adjusted by the leadership council without an RFC, I think it would be possible for them to make an immediate one-time exception if needed (and it may not be needed; it's possible I'm worrying for nothing). Do you agree?

Reviewer 5: I think that if we need to represent sponsor interests on the funding team, we should explicitly write that into the initial charter, either permanently or with a time bound. We could always make that separate from the project part of the team and its affiliation limits. That aside, it would not be too hard to have a team just barely large enough to have two members per affiliation. But if neither of those things suffice, then yes, I can imagine a world in which we make a careful exception, perhaps time-bounded. (We would also need to be mindful of how that looks.)

Reviewer 9: I don't see why funders should be represented on the funding team. That sounds like a clear conflict of interest to me.

Author: I will add Josh's text, which I think is a reasonable starting off point. I don't see the conflict-of-interest that @Kobzol mentioned myself, but I also don't really know what "represents sponsors interest" means to those speaking. My feeling is that large sponsors likely don't even *want* to be "on the team", it's a lot of work with very low ROI. They probably do want to be able to put in their two cents about what areas to fund and be kept in the loop, that kind of thing, but that sounds like a standard conversation to be had with the sponsor.

## The Funding team's day-to-day responsibilities

The Funding team's role is to keep a pulse on the project and work with the Foundation to select which maintainers to hire. Its core responsibilities are:

1. **Staying in regular contact with teams** — meeting regularly with team leads and members to understand their needs, health, and where support would have the most impact.
    * These meetings can be helpful for purposes other than hiring a MiR: the Funding team may be able to connect the team with Foundation resources to resolve a situation.
2. **Working with the Foundation to select MiR candidates** — when positions become available, evaluating applicants and selecting candidates who'll have the most overall impact based on project needs.
3. **Collecting feedback on how well the program and the MiRs are working** — as the team responsible for selecting which maintainers to hire, the Funding team is also the team responsible for fielding feedback on how well those decisions work out and making adjustments as needed.

The Foundation supports the Funding team with logistics. The Foundation issues contracts or manages employment.

> The Leadership Council appoints new members for the Funding team, for fixed terms (e.g. one year at a time), which can be renewed. This helps ensure that the membership reflects the will and needs of broader Project teams, since the Funding team should consider the needs of all of those teams. The charter of the Funding team should capture this.

Reviewer 10: *[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3931/c143e315774f746d667c5eecd95a8ed999e8a729..7a84a3585868d2def0a1a7fbca7779d087fa2ef9)*

```suggestion
To ensure the membership reflects the will and needs of the teams of the Project overall, new members will be appointed to the Funding team for fixed terms (e.g., one year at a time) which can be renewed.  Either the council will make these appointments or the members will be directly appointed by the top-level teams.  This will be decided when the council adopts the team charter.
```

I had reviewed the draft ahead of the addition of this line.  I think it's just a bit too specific.  One of the reasons I put https://github.com/rust-lang/rfcs/pull/3919 into draft was to adjust it for direct appointment by teams rather than requiring LC appointment.  There are merits to each, but I came to feel it'd be more representative for the teams to directly appoint rather than requiring the LC to make a consensus decision about this.  And the cost of the LC having to make these decisions is rather high.

I don't think we need to decide this now; we can do it when adopting the charter.  But I'd like to leave this door open.

Reviewer 9: I don't really have any strong opinion on this, but I was curious how would you imagine it to work for the teams to appoint people to some team? That all members of a team would vote to select one of their members (or anyone from the Project) to be on the team? That seems tricky for t-devtools and especially t-launching-pad, which represent a large amount of teams in the Council; I suppose that the fact that there is a council representative was also kind of created to deal with this.

Reviewer 5: > I don't really have any strong opinion on this, but I was curious how would you imagine it to work for the teams to appoint people to some team?

I would suggest *for now* that we consider just having the LC appoint the initial team, to get things started. We could revise later easily enough.

The Leadership Council has a Project Priorities budget, which is used to fund various initiatives, such as travel grants or program management. RFMF funds will be directed to this budget, but they will be dedicated to activities that direct funding to individual maintainers. This includes the existing program management program and would include the proposed Project Grants Program ([RFC 3919]), which provides modest stipends to recognize and support existing contributors. It will also include a third program, the *Maintainer in Residence program*, proposed by this RFC.

The Maintainer in Residence program is dedicated to hiring long-term maintainers and funding their maintenance work in full. Maintainers' in Residence time is split between priorities guided by the teams they are supporting and priorities of their own choosing within the project.

Selecting Maintainers in Residence is a collaboration between the Foundation and a "Funding team" appointed by the Leadership Council. This Funding team will weigh the set of applications against the project's needs and priorities.

> Funding team members must not disproportionately come from any one company, legal entity, or closely related set of legal entities, to avoid impropriety or the appearance of impropriety. If the Funding team has 5 or fewer representatives, no more than 1 representative may have any given affiliation; if the Funding has 6 or more representatives, no more than 2 representatives may have any given affiliation. (This affiliation limit comes from [RFC 3392](https://rust-lang.github.io/rfcs/3392-leadership-council.html#limits-on-representatives-from-a-single-companyentity), and the Funding team charter should import the definition of affiliations and the processes around handling them from there. The Leadership Council may update this policy through its normal decision-making process.)

Reviewer 5: ```suggestion
Funding team members must not disproportionately come from any one company, legal entity, or closely related set of legal entities, to avoid impropriety or the appearance of impropriety. If the Funding team has 5 or fewer representatives, no more than 1 representative may have any given affiliation; if the Funding team has 6 or more representatives, no more than 2 representatives may have any given affiliation. (This affiliation limit comes from [RFC 3392](https://rust-lang.github.io/rfcs/3392-leadership-council.html#limits-on-representatives-from-a-single-companyentity), and the Funding team charter should import the definition of affiliations and the processes around handling them from there. The Leadership Council may update this policy through its normal decision-making process.)
```

Accidentally introduced a typo.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3931/c143e315774f746d667c5eecd95a8ed999e8a729..016041716243558a2c64b07f1ff2b7562554b9fd)*

Reviewer 12: Resolved by https://github.com/rust-lang/rfcs/pull/3931/commits/0b691454e95ef6ed43a07ade4b76d81514db2939

- Feature Name: `rust_foundation_maintainer_fund`
- Start Date: 2026-02-23
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)

## Summary
[summary]: #summary

This RFC defines the relationship between the Rust Foundation Maintainer Fund (RFMF) and the open-source Rust project. The RFMF is a dedicated fund used to support Rust maintenance: open-ended, multiplicative work that improves Rust and its codebase.

> The Leadership Council has a Project Priorities budget, which is used to fund various initiatives, such as travel grants or program management. RFMF funds will be directed to this budget, but they will be dedicated to activities that direct funding to individual maintainers. This includes the existing program management program and would include the proposed Project Grants Program ([RFC 3919]), which provides modest stipends to recognize and support existing contributors. It will also include a third program, the *Maintainer in Residence program*, proposed by this RFC.

Reviewer 4: ```suggestion
The Leadership Council has a Project Priorities budget, which is used to fund various initiatives, such as travel grants or program management. RFMF funds will be directed to this budget, but they will be dedicated to activities that direct funding to individual maintainers. This includes the existing program management program and would include the proposed Project Grants Program ([RFC 3919]), which provides modest stipends to recognize and support existing contributors. It will also include a third program, the *Maintainer in Residence (MiR) program*, proposed by this RFC.
```

The acronym `MiR` becomes clear as "Maintainers in Residence" only down this document, I would probably do that here (first occurrence, AFAICS)

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3931/c143e315774f746d667c5eecd95a8ed999e8a729..0b691454e95ef6ed43a07ade4b76d81514db2939)*
